android-record/src/main/java/com/fasterxml/jackson/module/androidrecord/AndroidRecordModule.java

```diff
@@ -0,0 +1,95 @@
+ * using the canonical constructor on Android,
+ * where java records are supported through desugaring,
+ * and Jackson's built-in support for records doesn't work,
+ * since the desugared classes have a non-standard super class,
+ * and record component-related reflection methods are missing.
+ *
+ * <p>
+ * See <a href="https://android-developers.googleblog.com/2023/06/records-in-android-studio-flamingo.html">
+ *   Android Developers Blog article</a>
+ *
+ * <p>
+ * Note: this module is a no-op when no Android-desugared records are being deserialized,
+ * so it is safe to use in code shared between Android and non-Android platforms.
+ *
+ * <p>
+ * Note: The canonical record constructor is found
+ * through matching of parameter types with field types.
+ * Therefore, this module doesn't allow a deserialized desugared record class to have a custom
+ * constructor with a signature that's any permutation of the canonical one's.
+ *
+ * @author Eran Leshem
+ **/
+public class AndroidRecordModule extends SimpleModule {
+  @Override
+  public void setupModule(SetupContext context) {
+    super.setupModule(context);
+    context.addValueInstantiators(AndroidRecordModule::findValueInstantiator);
+  }
+
+  private static ValueInstantiator findValueInstantiator(DeserializationConfig config, BeanDescription beanDesc,
+                                                         ValueInstantiator defaultInstantiator) {
+    Class<?> raw = beanDesc.getType().getRawClass();
+    if (defaultInstantiator instanceof StdValueInstantiator && raw.getSuperclass() != null
+            && raw.getSuperclass().getName().equals("com.android.tools.r8.RecordTag")) {
+      Map<Type, Integer> componentTypes = typeMap(Arrays.stream(raw.getDeclaredFields())
+              .filter(field -> !Modifier.isStatic(field.getModifiers())).map(Field::getGenericType));
+      boolean found = false;
+      for (Constructor<?> constructor: raw.getDeclaredConstructors()) {
+        Parameter[] parameters = constructor.getParameters();
+        Map<Type, Integer> parameterTypes = typeMap(Arrays.stream(parameters).map(Parameter::getParameterizedType));
+        if (! parameterTypes.equals(componentTypes)) {
+          continue;
+        }
+
+        if (found) {
+          throw new IllegalArgumentException(String.format(
+                          "Multiple constructors match set of components for record %s", raw.getName()));
+        }
+
+        SettableBeanProperty[] properties = new SettableBeanProperty[parameters.length];
+        for (int i = 0; i < parameters.length; i++) {
+          Parameter parameter = parameters[i];
+          properties[i] = CreatorProperty.construct(PropertyName.construct(parameter.getName()), config.getTypeFactory()
+                          .constructType(parameter.getParameterizedType()), null, null, null, null, i, null, null);
+        }
+
+        ((StdValueInstantiator) defaultInstantiator).configureFromObjectSettings(null, null, null, null,
```

Comment:
This looks dangerous... having mostly `null`s, and overwriting internal state.
Ideally I'd rather create a new instance (even if of type `StdValueInstantiator`) instead of forcibly trying to modify passed-in instantiator

Reply:
I've improved the Property construction code. Please take a look.
Regarding modifying the instantiator, I followed [this example](https://github.com/FasterXML/jackson-datatype-jsr310/blob/master/src/main/java/com/fasterxml/jackson/datatype/jsr310/JavaTimeModule.java#L232). Should I use a different pattern?

Reply:
Will do. I think that usage pattern is probably fine -- will add a comment if I remember reasons to use different approach.